src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -67,434 +113,658 @@ public JsonWebToken(string header, string payload)
+#else
+                    Header = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, 0, Dot1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, encodedJson.Substring(0, Dot1), encodedJson), ex));
+                }
+
+                try
+                {
+#if NET45
+                    MessageBytes = Encoding.UTF8.GetBytes(encodedJson.ToCharArray(0, Dot2));
+                    _pChars = encodedJson.ToCharArray(Dot1 + 1, Dot2 - Dot1 - 1);
+                    Payload = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_pChars));
+#else
+                    Payload = new JsonClaimSet(JwtTokenUtilities.GetJsonDocumentFromBase64UrlEncodedString(encodedJson, Dot1 + 1, Dot2 - Dot1 - 1));
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14101, encodedJson.Substring(Dot2, Dot2 - Dot1), encodedJson), ex));
+                }
             }
-            catch (Exception ex)
+            else
             {
-                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, tokenParts[0], rawData), ex));
+                // empty payload for JWE's {encrypted tokens}.
+#if NET45
+                Payload = new JsonClaimSet("{}");
+#else
+                Payload = new JsonClaimSet(JsonDocument.Parse("{}"));
+#endif
+                if (Dot3 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                Dot4 = encodedJson.IndexOf('.', Dot3 + 1);
+
+                // JWE needs to have 4 dots
+                if (Dot4 == -1 || Dot4 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                // too many dots...
+                if (encodedJson.IndexOf('.', Dot4 + 1) != -1)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                // right number of dots for JWE
+                _hChars = encodedJson.ToCharArray(0, Dot1);
+
+                // header cannot be empty
+                if (_hChars.Length == 0)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14307, encodedJson)));
+
+                HeaderAsciiBytes = Encoding.ASCII.GetBytes(_hChars);
+                try
+                {
+                    Header = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_hChars));
+                }
+                catch (Exception ex)
```

Comment:
>Exception [](http://example.com/codeflow?start=23&length=9)

Are there cases where this isn't a FormatException or JsonReaderException? Otherwise it would be more unexpected I would think and may indicate something more serious than a bad argument.

Reply:
We can address this when we have a model that allows users to examine exceptions.
Given where we are in the stack, the next handler may succeed, so we don't want to throw back to the application.